src/modules/lua_mtev.c

```diff
@@ -3419,120 +3417,125 @@ nl_watchdog_child_heartbeat(lua_State *L) {
-    return 1;
   }
   udata = lua_touserdata(L, 1);
   node = *udata;
   if(!lua_isstring(L, 2)) {
     luaL_error(L, "metatable error, arg2 not a string!");
-    return 1;
   }
-
   k = lua_tostring(L, 2);
-  switch (*k) {
+  switch(*k) {
+    case 'b':
+      if(!strcmp(k, "boot_time")) {
+        mtev_lua_push_timeval(L, node->boot_time);
+      }
+      else break;
+      return 1;
     case 'i':
       if(!strcmp(k, "id")) {
         char uuid_str[UUID_PRINTABLE_STRING_LENGTH];
         uuid_unparse_lower(node->id, uuid_str);
         lua_pushstring(L, uuid_str);
-      } else if(!strcmp(k, "id_ud")) {
-        lua_pushlightuserdata(L, (void*)&node->id);
-      } else {
-        break;
       }
+      else break;
+      return 1;
+    case 'l':
+      if(!strcmp(k, "last_contact")) {
+        mtev_lua_push_timeval(L, node->last_contact);
+      }
+      else break;
       return 1;
     default:
       break;
   }
-  luaL_error(L, "mtev_cluster_node_t no such element: %s", k);
+  luaL_error(L, "noit_module_t no such element: %s", k);
   return 0;
 }
 
 static void
-mtev_lua_setup_cluster_node(lua_State *L,
-    mtev_cluster_node_t *cluster) {
+mtev_lua_setup_cluster_node(lua_State *L, mtev_cluster_node_t *node) {
   mtev_cluster_node_t **addr;
-  addr = (mtev_cluster_node_t **)lua_newuserdata(L, sizeof(cluster));
-  *addr = cluster;
-  if(luaL_newmetatable(L, "mtev_cluster_node_t") == 1) {
-    lua_pushcclosure(L, mtev_lua_cluster_node_index_func, 0);
-    lua_setfield(L, -2, "__index");
+  if(node == NULL) {
+    lua_pushnil(L);
+  } else {
+    addr = (mtev_cluster_node_t **)lua_newuserdata(L, sizeof(node));
+    *addr = node;
+    if(luaL_newmetatable(L, "mtev_cluster_node_t") == 1) {
+      lua_pushcclosure(L, mtev_lua_cluster_node_index_func, 0);
+      lua_setfield(L, -2, "__index");
+    }
+    lua_setmetatable(L, -2);
   }
-  lua_setmetatable(L, -2);
 }
 
-static mtev_hook_return_t
-mtev_lua_on_cluster_node_updated(void *closure, mtev_cluster_node_t *updated_node, mtev_cluster_t *cluster) {
-  lua_callback_ref *cb_ref;
-  int rv;
-  cb_ref = closure;
+static void
+mtev_lua_push_cluster_details(lua_State *L, mtev_cluster_t *cluster, mtev_cluster_node_t **nodes, int number_of_nodes) {
+  int i;
 
-  mtevAssert(pthread_equal(cb_ref->self, pthread_self()));
+  lua_createtable(L, 0, 2);
 
-  lua_rawgeti( cb_ref->L, LUA_REGISTRYINDEX, cb_ref->callback_reference );
-  lua_rawgeti( cb_ref->L, LUA_REGISTRYINDEX, cb_ref->closure_reference );
-  mtev_lua_setup_cluster_node(cb_ref->L, updated_node);
-  lua_pushlightuserdata(cb_ref->L, (void*)&cluster);
+  lua_pushstring(L, "oldest_node");
+  mtev_lua_setup_cluster_node(L, mtev_cluster_get_oldest_node(cluster));
+  lua_settable(L, -3);
 
-  lua_call(cb_ref->L, 3, 1);
-  rv = lua_tointeger(cb_ref->L, -1);
+  lua_pushstring(L, "nodes");
+  lua_createtable(L, number_of_nodes, 0);
+  for(i=0; i != number_of_nodes; ++i) {
+    lua_pushinteger(L, i+1);
+    mtev_lua_setup_cluster_node(L, nodes[i]);
```

Comment:
This node[i] could be freed after return but before use in lua. We need a "dumb lua" copy here.